src/catalog/database_catalog.cpp

```diff
@@ -953,11 +1049,1065 @@ const Schema &DatabaseCatalog::GetSchema(const common::ManagedPointer<transactio
+          }
+        }
+        verify_res = VerifyFKConstraint(txn, con_obj, pr);
+      }
+      else if (con_obj.contype_ == postgres::ConstraintType::CHECK){
+        // TODO: implement support for check constraint
+        con_obj.AddCheckConstraintMetaData(con_check);
+        verify_res = VerifyCheckConstraint(con_obj);
+      }
+      else if (con_obj.contype_ == postgres::ConstraintType::EXCLUSION){
+        // TODO: implement support for exclusion constraint
+        con_obj.AddExclusionConstraintMetadata(con_exclusion);
+        verify_res = VerifyExclusionConstraint(con_obj);
+      }
+
+    if (!verify_res) {
+      delete[] child_buffer;
+      delete[] buffer;
+      txn->SetMustAbort();
+      return false;
+    }
+  }
+  delete[] child_buffer;
+  delete[] buffer;
+  return true;
+}
+
+void DatabaseCatalog::CopyColumnData(storage::ProjectedRow *table_pr, storage::ProjectedRow *index_pr,
+                                     std::vector<col_oid_t> col_vec, const Schema &schema) {
+  auto col_offset_map = ColToOffsetMap(schema);
+  for (uint16_t col_index = 0; col_index < col_vec.size(); col_index ++) {
+    auto index_pr_index = col_index; // idx of the pr for index retrieval
+    auto table_pr_index = col_offset_map[col_vec[col_index]]; // index to get the data from pr
+    auto *const index_ptr = index_pr->AccessForceNotNull(index_pr_index);
+    auto *const pr_ptr = table_pr->AccessForceNotNull(table_pr_index);
+    const auto &table_col = schema.GetColumn(table_pr_index);
+    if (table_col.Type() == type::TypeId::INTEGER) {
+      col_oid_t val = *(reinterpret_cast<col_oid_t *>(pr_ptr));
+      *(reinterpret_cast<col_oid_t *>(index_ptr)) = val;
+    }
+    else if (table_col.Type() == type::TypeId::VARCHAR || table_col.Type() == type::TypeId::VARBINARY) {
+      std::memcpy(index_ptr, pr_ptr, table_col.MaxVarlenSize());
+    } else {
+      std::memcpy(index_ptr, pr_ptr, type::TypeUtil::GetTypeSize(table_col.Type()));
+    }
+  }
+}
+
+bool DatabaseCatalog::VerifyUniquePKConstraint(common::ManagedPointer<transaction::TransactionContext> txn,
+    const PG_Constraint &con_obj,
+    storage::ProjectedRow *pr) {
+  const Schema &table_schema = GetSchema(txn, con_obj.conrelid_);
+  TERRIER_ASSERT(con_obj.concol_.size() > 0, "UNIQUE and PK constraint should always have affecting column");
+  common::ManagedPointer<storage::index::Index> index = GetIndex(txn, con_obj.conindid_);
+  const auto pri = index->GetProjectedRowInitializer();
+  auto *const buffer = common::AllocationUtil::AllocateAligned(pri.ProjectedRowSize());
+  auto *key_pr = pri.InitializeRow(buffer);
+  CopyColumnData(pr, key_pr, con_obj.concol_, table_schema);
+  std::vector<storage::TupleSlot> index_scan_results;
+  index->ScanKey(*txn, *key_pr, &index_scan_results);
+  // set the index projected row from source projected row
+  if (!index_scan_results.empty()) {
+    delete[] buffer;
+    return false;
+  }
```

Comment:
Do you want to check when result is greater than 1 since it's checking for uniqueness?